src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -91,19 +91,19 @@ private static SocketAsyncEngine[] CreateEngines()
         //
         private readonly ConcurrentQueue<SocketIOEvent> _eventQueue = new ConcurrentQueue<SocketIOEvent>();
 
-        // The scheme works as follows:
-        // - From NotScheduled, the only transition is to Scheduled when new events are enqueued and a work item is enqueued to process them.
-        // - From Scheduled, the only transition is to Determining right before trying to dequeue an event.
-        // - From Determining, it can go to either NotScheduled when no events are present in the queue (the previous work item processed all of them)
-        //   or Scheduled if the queue is still not empty (let the current work item handle parallelization as convinient).
-        //
-        // The goal is to avoid enqueueing more work items than necessary, while still ensuring that all events are processed.
-        // Another work item isn't enqueued to the thread pool hastily while the state is Determining,
-        // instead the parallelizer takes care of that. We also ensure that only one thread can be parallelizing at any time.
         private enum EventQueueProcessingStage
         {
+            // NotScheduled: has no guarantees
             NotScheduled,
-            Determining,
+
+            // Scheduled: means a worker will check work queues and ensure that
+            // any work items inserted in work queue before setting the flag
+            // are picked up.
+            // Note: The state must be cleared by the worker thread _before_
+            //       checking. Otherwise there is a window between finding no work
+            //       and resetting the flag, when the flag is in a wrong state.
+            //       A new work item may be added right before the flag is reset
+            //       without asking for a worker, while the last worker is quitting.
             Scheduled
         }
```

Comment:
With just two not-scheduled / scheduled values now, do we even still need the enum? Could it just be a `scheduled` bool?

Reply:
I've replaced the `enum` with an `int`, as it was before. A `bool` turned out somewhat inconvenient:
- We do Interlocked operations with this variable and by looking at the implementation interlocked with bool may have less optimal than int implementation, depending on platform. Here we want to be fairly efficient.
- In one case the variable is padded into a cacheline-sized struct and there is a pattern used to pad `int32` fields that is used in many places, but not for `bool`.